### ISSUE TITLE
Invalid SSL certificate file.

### ISSUE THREAD

**Issue by betam4x** [excon/excon#473]
When using Excon I get the following: 

    Excon::Errors::SocketError: Unable to verify certificate, please set `Excon.defaults[:ssl_ca_path] = path_to_certs`, `ENV['SSL_CERT_DIR'] = path_to_certs`, `Excon.defaults[:ssl_ca_file] = path_to_file`, `ENV['SSL_CERT_FILE'] = path_to_file`, `Excon.defaults[:ssl_verify_callback] = callback` (see OpenSSL::SSL::SSLContext#verify_callback), or `Excon.defaults[:ssl_verify_peer] = false` (less secure).

Working on a Windows 10 machine, but i've reproduced in a couple Windows 7 environments, haven't tried a mac or linux yet.  the SSL_CERT_FILE is set properly and mucking with the defaults for Excon does nothing.  Only disabling verification altogether works.  Also, other libraries I've used (that use OpenSSL) work fine, so it appears to be specific to Excon.

    irb(main):004:0> puts ENV['SSL_CERT_FILE']
    C:\Ruby21-x64\cacert.pem

**Comment by geemus** [comment]
Were other libraries you used also in Ruby or other stuff? Could you point out specific ones? I might be able to spelunk in their code to figure out what might be different. Otherwise I'm afraid I don't have much experience on the windows side of things.

**Comment by johngtimms** [comment]
Hi. I am experiencing the same issue on Windows 7, so hopefully I can provide some specifics. I have a team of people working on a Rails project and we just set up environments so this issue is very fresh in my memory. 

When you install Ruby 2.1 on Windows (using [RubyInstaller](http://rubyinstaller.org/)) and try to install gems (say, like `gem install bundler`) you will get an SSL error. I don't have the exact error in front of me, but following the steps [here](https://gist.github.com/fnichol/867550) fixes the problem and you can then use `gem`. Notice that the solution involves setting the `SSL_CERT_FILE` environment variable.

The Excon error that everyone gets says to set that same environment variable, but most people on Windows already have that variable set (assuming they can install gems!). I can confirm the report from @betam4x :

    Microsoft Windows [Version 6.1.7601]
    Copyright (c) 2009 Microsoft Corporation.  All rights reserved.
    
    C:\Users\John>irb
    DL is deprecated, please use Fiddle
    irb(main):001:0> puts ENV["SSL_CERT_FILE"]
    C:\Ruby\ca-bundle.crt
    => nil
    irb(main):002:0>

Please let me know what other debug info I can provide. I've got a team of developers that's trying to work around this issue now. We are getting the error when using Carrierwave with Fog to talk to AWS S3.

tl;dr Excon isn't reading ENV["SSL_CERT_FILE"] on Windows. RubyGems is an example of a program that works correctly.

**Comment by betam4x** [comment]
John:  thanks for that.

Everyone: sorry for the lack of replies.  I meant to get back to you, but I
have been busy.

**Comment by geemus** [comment]
Any chance SSL_CERT_DIR would also be set? Also, any ssl related settings being passed to carrierwave? I'm just wondering if some other value is inadvertently stomping on this one, even though it is set.

**Comment by johngtimms** [comment]
No. I tested all of the variables/defaults listed in the error, none of them are set.

I'm not experienced in Ruby debugging, but I made [a short script](https://gist.github.com/johngtimms/0c42c687b1c0e9022cbf) that uses Carrierwave to upload a single file to S3, causing an error. I hope it's helpful. (Notice that the trace shows the error arising from something Fog is doing, not Carrierwave directly.) If I have time, I will try to put together a script that just uses Fog and still gets an error.

**Comment by geemus** [comment]
Thanks, could you actually change the puts to use `puts example.inspect`? I think whether they are nil vs empty string could actually be an issue in this case, so I'd like to be certain. Thanks!

**Comment by johngtimms** [comment]
On Windows:

    puts Excon.defaults[:ssl_ca_path].inspect           => nil
    puts ENV['SSL_CERT_DIR'].inspect                    => nil
    puts Excon.defaults[:ssl_ca_file].inspect           => nil
    puts ENV['SSL_CERT_FILE'].inspect                   => "C:\\Ruby\\ca-bundle.crt"
    puts Excon.defaults[:ssl_verify_callback].inspect   => nil
    puts Excon.defaults[:ssl_verify_peer].inspect       => true

I ran the same tests on Ubuntu (where Excon works perfectly) and interestingly enough the results were the same **except** `puts ENV['SSL_CERT_FILE'].inspect` was also `nil`.

**Comment by geemus** [comment]
Thanks, I suspected this might not gain us much, but always good to be certain.

**Comment by geemus** [comment]
You wouldn't happen to be working through a proxy in the windows case, would you?

**Comment by betam4x** [comment]
I get basically the same result on windows (with the exception of the SSL_CERT_FILE being a different filename) and I'm not behind a proxy.

**Comment by johngtimms** [comment]
Nope, no proxy.

**Comment by geemus** [comment]
Some others are seeing some issues related to a more recently released openssl. What version are people using here? I'm not sure it is related, but, again, trying to reduce variables. Thanks!

**Comment by johngtimms** [comment]
I'm not terribly familiar with how openssl works on Windows. I ran `ruby -ropenssl -e 'puts OpenSSL::OPENSSL_VERSION'` and got `OpenSSL 1.0.0o 15 Oct 2014`. Is that helpful?

I don't use Cygwin, and I installed Ruby using RubyInstaller, but I do have Git for Windows, and I'm not sure what it does to my path/environment.

**Comment by geemus** [comment]
This looks relevant: https://groups.google.com/forum/#!msg/rubyinstaller/DVIDvs7xKC0/8dqXCqmABk0J

It seems to suggest that `SSL_CERT_FILE` may be the one we need/want to change (and that this is indeed a railsinstaller specific issue).

**Comment by johngtimms** [comment]
Hmm, I thought we might be getting somewhere, but I don't believe this is it. On Windows:

    puts OpenSSL::X509::DEFAULT_CERT_FILE    => C:/Users/Luis/Code/luislavena/knap-build/var/knapsack/software/x64-windows/openssl/1.0.0l/ssl/cert.pem

On Ubuntu:

    puts OpenSSL::X509::DEFAULT_CERT_FILE    => /usr/lib/ssl/cert.pem

Neither path exists, so this isn't what's causing the problem. However, I think you might be going in the right direction- Excon is maybe getting a bad path from somewhere other than the settings mentioned in the error message. That somewhere just isn't `OpenSSL::X509::DEFAULT_CERT_FILE`.

The way I read the Groups message is that Davis is requesting that the RailsInstaller team set `SSL_CERT_FILE` when their installer runs. RailsInstaller still does not do that, so me and @betam4x have set this variable manually.

**Comment by johngtimms** [comment]
OK, here we go. On Windows:

    puts OpenSSL::X509::DEFAULT_CERT_DIR    => C:/Users/Luis/Code/luislavena/knap-build/var/knapsack/software/x64-windows/openssl/1.0.0o/ssl/certs

On Ubuntu:

    puts OpenSSL::X509::DEFAULT_CERT_DIR    => /usr/lib/ssl/certs

On Windows, the path does not exist. On Ubuntu, it does, and it's full of certificates. I tried creating the path in Windows, putting certificates there, and running the test script I provided- I still got the error. (It's possible I wasn't using the right certificates format?) On Ubuntu, I tried renaming the `certs` directory, ran the script, and I got the error. I set it back to the original name and the script ran fine.

I think it's safe to conclude that Excon is relying on `OpenSSL::X509::DEFAULT_CERT_DIR` and is preferring that path over the paths named in the error, at least on Windows. I have a few other ideas to try (What would happen if I set `ENV['SSL_CERT_FILE']` on Ubuntu and moved `/usr/lib/ssl/certs`? What would happen if I copied all `/usr/lib/ssl/certs/*` to the `C:/Users/Luis/.....` folder on Windows?) but I don't have the time today.

**Comment by clarkdave** [comment]
I'm also having this issue. Excon works fine on my Mac but as soon as I deploy to our Ubuntu 12.04 servers SSL connections just will not work.

Using excon `0.43.0` with Faraday `0.91`. The exact error is:

```
Faraday::SSLError: Unable to verify certificate, please set `Excon.defaults[:ssl_ca_path] = path_to_certs`, `ENV['SSL_CERT_DIR'] = path_to_certs`, `Excon.defaults[:ssl_ca_file] = path_to_file`, `ENV['SSL_CERT_FILE'] = path_to_file`, `Excon.defaults[:ssl_verify_callback] = callback` (see OpenSSL::SSL::SSLContext#verify_callback), or `Excon.defaults[:ssl_verify_peer] = false` (less secure).
```

I've tried setting `ENV['SSL_CERT_DIR'] = /etc/ssl/certs`, etc. Nothing seems to work except `[:ssl_verify_peer] = false`.

I just cannot figure it out - `Net::HTTP` can make the same https request without errors, as can `cURL`, etc.

Happy to give more info if it helps!

**Comment by johngtimms** [comment]
@clarkdave I'm not the developer, but could you report the output of the following in `irb`?

    require 'openssl'
    puts OpenSSL::X509::DEFAULT_CERT_FILE
    puts OpenSSL::X509::DEFAULT_CERT_DIR

If those paths aren't `nil`, is there anything at the location(s) they point to? This is the first I've heard of this issue appearing on Ubuntu after `ENV['SSL_CERT_DIR']` (or similar) is set.

**Comment by clarkdave** [comment]
@johngtimms OK I must be going out of my mind because I've just woken up and tried this again and it works! I rolled back all my fixes (`ENV['SSL_CERT_DIR']` etc) but can't replicate the error anymore.

Sorry in advance to any future fellow problem-havers! Try going to bed. It might magically fix itself by the following morning :zzz: 

For posterity:

```
irb(main):001:0> require 'openssl'
=> false
irb(main):002:0> puts OpenSSL::X509::DEFAULT_CERT_FILE
/usr/lib/ssl/cert.pem
=> nil
irb(main):003:0> puts OpenSSL::X509::DEFAULT_CERT_DIR
/usr/lib/ssl/certs
=> nil
```

`/usr/lib/ssl/cert.pem` doesn't exist, but `/usr/lib/ssl/certs` does and is a symlink to `/etc/ssl/certs`, which is full of certificates.

**Comment by jessethebuilder** [comment]
I get this issue when I try to use carrierwave and s3. Is there a fix?

**Comment by geemus** [comment]
@jessethebuilder with no specified ssl stuff it should fallback to a reasonable default. Otherwise, could you share what you are setting and what you are seeing? Thanks!

**Comment by vsolari** [comment]
I had the same problem when trying to integrate "Carrierwave", "Fog" and "Amazon S3", the solution was to use an older "excon" gem, so I had to uninstall "excon 0.4XX" and use "excon 0.39.5 "

They must be synchronized date and time of your server to upload documents to "Amazon S3"

Sorry for my English :)

**Comment by hkmaly** [comment]
I had the same problem with Fog, Excon and rackspace. I noticed (using strace) that the program tries to open file /usr/lib/ssl/certs/98ec67f0.0 which is a symlink to ca-certificates.crt, big file with multiple certificates. On different server, the file was symlink to Thawte_Premium_Server_CA.pem ; I changed the symlink and it started to work.

Debian wheezy, relevant packages:
libssl1.0.0/wheezy uptodate 1.0.1e-2+deb7u17
openssl/wheezy uptodate 1.0.1e-2+deb7u17
ssl-cert/wheezy uptodate 1.0.32+deb7u1
ca-certificates/wheezy uptodate 20130119+deb7u1

excon (0.45.4)

I suspect that excon have some problem reading that long file ...

**Comment by geemus** [comment]
Thanks for the detailed report!

**Comment by sandstrom** [comment]
@geemus I've run into a similar problem. It took me a while to understand the error message. Perhaps it's just me, but I think it could be reworded for slightly more clarity.

What do you think about something like this?

"Unable to verify certificate. This may be an issue with the remote host, or with Excon. Excon has a bundled set of certificates, but these can be customized. Set `Excon.defaults[:ssl_ca_path] = path_to_certs`, `ENV['SSL_CERT_DIR'] = path_to_certs`, `Excon.defaults[:ssl_ca_file] = path_to_file`, `ENV['SSL_CERT_FILE'] = path_to_file`, `Excon.defaults[:ssl_verify_callback] = callback` (see OpenSSL::SSL::SSLContext#verify_callback), or `Excon.defaults[:ssl_verify_peer] = false` (less secure)."

https://github.com/excon/excon/blob/75d85a7e304cbd1c9dc3c7c40c6de5a995f5cd04/lib/excon/errors.rb#L13

**Comment by geemus** [comment]
@sandstrom Yeah, sounds good. Would you be up for making that change? (might be nice to tackle: https://github.com/excon/excon/issues/513 at the same time. Basically adding a more specific error, probably something like Excon::Errors::BadCertificateError)

**Comment by geemus** [comment]
Closing this issue due to relative inactivity. @sandstrom I think the wording tweaks you suggest would still be solid, I just don't have enough to go on toward a broader fix from the discussion above.
